typing/typecore.ml

```diff
@@ -5089,6 +5139,49 @@ and type_andops env sarg sands expected_ty =
   let let_arg, rev_ands = loop env sarg (List.rev sands) expected_ty in
   let_arg, List.rev rev_ands
 
+(*TODO mbungeroth: Why do I need the type annotation here to make this work?*)
+  and comprehension ~loc ~env ~ty_expected ~container_type 
+                  ~sbody ~(comp_typell : Parsetree.comprehension list list) = 
+
+  let from_to_comprehension ~body_env ~env param slow shigh dir =
+    let low = type_expect env slow
+        (mk_expected ~explanation:For_loop_start_index Predef.type_int) in
+    let high = type_expect env shigh
+        (mk_expected ~explanation:For_loop_stop_index Predef.type_int) in
+    let id, new_env = type_for_loop ~loc ~env:body_env ~param  Predef.type_int in
+    (*let body = type_expect new_env sbody ty_expected in*)
+    id, From_to(param, low, high, dir), new_env
+  in
+  let in_comprehension ~body_env ~env param siter= 
+    let item_ty = newvar() in
+    let iter_ty = instance (container_type item_ty) in
+    let iter = type_expect env siter 
+        (mk_expected ~explanation:InComprehensionArgument iter_ty) in
+    let id, new_env = type_for_loop ~loc ~env:body_env ~param item_ty
+    in
+    id, In(param, iter), new_env
+  in
+  let call (ids, comps, body_env) ~env ~(comp_type : Parsetree.comprehension) = 
+    let id, comp, env = match comp_type with 
+    | From_to (p,e2,e3, dir) -> from_to_comprehension ~body_env ~env p e2 e3 dir
+    | In (p, e2) ->  in_comprehension ~body_env ~env p e2 in
+    id::ids, comp::comps, env in 
+ 
+  let ids, comps, new_env = List.fold_left 
+    (fun (ids, comps, env) comp_typel -> 
+        let new_ids, new_comps, new_env  = 
+            List.fold_left 
+                (fun acc comp_type -> call acc ~env ~comp_type ) 
+                ([], [], env)  
+                comp_typel 
+            in
+            new_ids::ids, new_comps::comps, new_env
+    ) 
+    ([], [], env) comp_typell
+  in
+  let body = type_expect new_env sbody ty_expected in
```

Comment:
I'd be tempted to split the typing of the body out of this function, and have this one return `ids, env, comps`. That way you can avoid needing to pass in `ty_expected` and `sbody`.